test/testdata/desugar/pin_mlhs.rb

```diff
@@ -0,0 +1,16 @@
+# typed: true
+extend T::Sig
+
+sig { returns([Integer, String]) }
+def tuple_int_str = [0, '']
+
+sig { returns([String, Integer]) }
+def tuple_str_int = ['', 0]
+
+def mlhs_example
+  x, _ = tuple_str_int
+  2.times do
+    y, _ = tuple_int_str
+    #  ^ error: Changing the type of a variable is not permitted
```

Comment:
This doesn't seem like the right error: we should be creating a new variable here, shouldn't we?  We would if the variable was actually named, right?

I can believe that the locs are better with the change, but I am a bit skeptical that this is the right way to demonstrate it.  But I am also on my phone and might not be thinking clearly.

Reply:
The `_` is not special in Ruby in unless you're using pattern matching syntax, which this is not. In this situation, `_` is just a variable name, like `x` or `y`. In this case, the `_` variable had type `Integer` outside the loop and has type `String` after this line runs inside the loop, and that's enough for Ruby to not like this code.

We could separately enforce that variables that start with underscores cannot be used (or maybe just that the `_` variable cannot be used) but that's tangential.

We could possibly change the logic which computes blocks arguments to say that if a variable is written to before it's read to in a block that it doesn't need to be in the basic block's argument list and then maybe doesn't participate in pinning, but that's a separate issue—the issue at hand is which location we use to report the pinning error that's happening.

Reply:
Sorry, you're right.  I think I was thinking more in Rust terms than in Ruby terms, as though the multi assignment somehow declared an entirely new set of variables.